plugins/crypto/mbedtls/ua_securitypolicy_aes128sha256rsaoaep.c

```diff
@@ -642,6 +642,144 @@ updateCertificateAndPrivateKey_sp_aes128sha256rsaoaep(UA_SecurityPolicy *securit
+
+    /* Set key usage in CSR context */
+    if(mbedtls_x509write_csr_set_key_usage(&request, MBEDTLS_X509_KU_DIGITAL_SIGNATURE |
+                                                     MBEDTLS_X509_KU_DATA_ENCIPHERMENT |
+                                                     MBEDTLS_X509_KU_NON_REPUDIATION |
+                                                     MBEDTLS_X509_KU_KEY_ENCIPHERMENT) != 0) {
+        mbedtls_x509_crt_free(&x509Cert);
+        mbedtls_x509write_csr_free(&request);
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    /* Add entropy */
+    if(nonce != NULL && nonce->length > 0) {
+        if (mbedtls_entropy_update_manual(&pc->entropyContext, (const unsigned char*)(nonce->data),
+                                          nonce->length) != 0) {
+            mbedtls_x509_crt_free(&x509Cert);
+            mbedtls_x509write_csr_free(&request);
+            return UA_STATUSCODE_BADINTERNALERROR;
+        }
+    }
+
+    /* Get subject from argument or read it from certificate */
+    char *subj = NULL;
+    if(subjectName != NULL && subjectName->length > 0) {
+        /* subject from argument */
+        subj = (char *)UA_malloc(subjectName->length + 1);
+        if(subj == NULL) {
+            mbedtls_x509_crt_free(&x509Cert);
+            mbedtls_x509write_csr_free(&request);
+            return UA_STATUSCODE_BADOUTOFMEMORY;
+        }
+        memset(subj, 0x00, subjectName->length + 1);
+        strncpy(subj, (char *)subjectName->data, subjectName->length);
+        /* search for / in subject and replace it by comma */
+        char *p = subj;
+        for(size_t i = 0; i < subjectName->length; i++) {
+            if (*p == '/' ) {
+                *p = ',';
+            }
+            ++p;
+        }
+    }
+    else {
+        /* read subject from certificate */
+        const size_t subjectMaxSize = 512;
+
+        mbedtls_x509_name s = x509Cert.subject;
+        subj = (char *)UA_malloc(subjectMaxSize);
+        if (subj == NULL) {
+            mbedtls_x509_crt_free(&x509Cert);
+            mbedtls_x509write_csr_free(&request);
+            return UA_STATUSCODE_BADOUTOFMEMORY;
+        }
+        ret = mbedtls_x509_dn_gets(subj, subjectMaxSize, &s);
+        if (ret <= 0) {
+            mbedtls_x509write_csr_free(&request);
+            UA_free(subj);
+            return UA_STATUSCODE_BADINTERNALERROR;
+        }
+    }
+
+    /* Set the subject in CSR context */
+    ret = mbedtls_x509write_csr_set_subject_name(&request, subj);
+    if(ret != 0) {
+        if (ret != 0) {
+            mbedtls_x509_crt_free(&x509Cert);
+            mbedtls_x509write_csr_free(&request);
+            UA_free(subj);
+            return UA_STATUSCODE_BADINTERNALERROR;
+        }
+    }
+
+    /* Get the subject alternate names from certificate and set them in CSR context*/
+    const mbedtls_x509_sequence *san_list = &x509Cert.subject_alt_names;
+    mbedtls_x509write_csr_set_subject_alt_name(&request, san_list);
```

Comment:
Copying the subject alternative names here is a good thing so far. The question is how did they initially came into the certificate? And how to handle the case, where the subject name shall remain unchanged but any subject alternative name (e.g. IP or DNS) must be changed?

Usually, when setting up a new device some kind of self-signed certificate comes into play. After setting the device up and configuring it properly, it is now the time to create a "final" certificate. This usually involves placing IP address and/or hostname of the device into the certificate. How can this be done when creating the certificate signing request?

Reply:
[The specification](https://reference.opcfoundation.org/Core/Part6/v105/docs/6.2.2) is a little vague about this topic:

> [Servers](https://reference.opcfoundation.org/search/359?t=Servers) shall specify a partial or a fully qualified dNSName [or a ](https://reference.opcfoundation.org/search/359?t=or%20a)static IPAddress which identifies the machine where the application [Instance](https://reference.opcfoundation.org/search/359?t=Instance) runs.